From 1a51a3a2de56842c775459299704ae8a2a7fa985 Mon Sep 17 00:00:00 2001
From: denglei <denglei@eswincomputing.com>
Date: Mon, 23 Dec 2024 20:17:49 +0800
Subject: [PATCH 351/652] WIN2030-16797:fix:Resolve the issue of abnormal stack
 in audio.

Changelogs:
	Resolve the issue of abnormal stack in audio.

Change-Id: Ie478e5b0c9c44c8967aa305aae8b4a4e7df6a872
Signed-off-by: denglei <denglei@eswincomputing.com>
---
 sound/soc/eswin/esw-dma.c       | 62 ++++++++++++++++++++-------------
 sound/soc/eswin/esw-i2s.c       | 16 +++++++--
 sound/soc/sof/eswin/Kconfig     |  1 +
 sound/soc/sof/eswin/Makefile    |  1 -
 sound/soc/sof/eswin/es-common.h |  2 +-
 5 files changed, 53 insertions(+), 29 deletions(-)

diff --git a/sound/soc/eswin/esw-dma.c b/sound/soc/eswin/esw-dma.c
index cb7bbfca3de51..387974d162706 100644
--- a/sound/soc/eswin/esw-dma.c
+++ b/sound/soc/eswin/esw-dma.c
@@ -159,11 +159,6 @@ int esw_pcm_dma_open(struct snd_soc_component *component,
 	substream->runtime->hw = hw;
 	substream->runtime->private_data = chip;
 
-	chip->conv_buf[substream->stream] = kzalloc(hw.period_bytes_max, GFP_KERNEL);
-	if (!chip->conv_buf[substream->stream]) {
-		return -ENOMEM;
-	}
-
 	return 0;
 }
 
@@ -256,13 +251,9 @@ int esw_pcm_dma_close(struct snd_soc_component *component,
 {
 	struct i2s_dev *chip = dev_get_drvdata(component->dev);
 
-	dmaengine_synchronize(chip->chan[substream->stream]);
-
 	dev_dbg(chip->dev, "%s\n", __func__);
 
-	if (chip->conv_buf[substream->stream]) {
-		kfree(chip->conv_buf[substream->stream]);
-	}
+	dmaengine_synchronize(chip->chan[substream->stream]);
 
 	return 0;
 }
@@ -425,49 +416,70 @@ int esw_pcm_dma_dai_register(struct i2s_dev *chip)
 {
 	struct dma_chan *chan0;
 	struct dma_chan *chan1;
-	const char *channel_names0 = "rx";
-	const char *channel_names1 = "tx";
+	const char *channel_names0 = "tx";
+	const char *channel_names1 = "rx";
 	int ret;
+	u32 period_bytes_max;
 
 	dev_dbg(chip->dev, "%s\n", __func__);
 
 	chan0 = dma_request_chan(chip->dev, channel_names0);
 	if (IS_ERR(chan0)) {
 		if (PTR_ERR(chan0) == -EPROBE_DEFER) {
-			dev_err(chip->dev, "request dma channel[%d] failed\n", SNDRV_PCM_STREAM_CAPTURE);
+			dev_err(chip->dev, "request dma channel[%s] failed\n", channel_names0);
 			return -EPROBE_DEFER;
 		}
-		dev_err(chip->dev, "dma channel[%d] is NULL\n", SNDRV_PCM_STREAM_CAPTURE);
+		dev_err(chip->dev, "dma channel[%s] is NULL\n", channel_names0);
 	} else {
-		chip->chan[SNDRV_PCM_STREAM_CAPTURE] = chan0;
+		chip->chan[SNDRV_PCM_STREAM_PLAYBACK] = chan0;
+		period_bytes_max = dma_get_max_seg_size(chan0->device->dev);
+		chip->conv_buf[SNDRV_PCM_STREAM_PLAYBACK] = kzalloc(period_bytes_max, GFP_KERNEL);
+		if (!chip->conv_buf[SNDRV_PCM_STREAM_PLAYBACK]) {
+			ret = -ENOMEM;
+			dev_err(chip->dev, "alloc conv buf0 err:%d\n", ret);
+			goto release_chan0;
+		}
 	}
 
 	chan1 = dma_request_chan(chip->dev, channel_names1);
 	if (IS_ERR(chan1)) {
 		if (PTR_ERR(chan1) == -EPROBE_DEFER) {
-			dev_err(chip->dev, "request dma channel[%d] failed\n", SNDRV_PCM_STREAM_PLAYBACK);
-			dma_release_channel(chip->chan[1]);
-			return -EPROBE_DEFER;
+			dev_err(chip->dev, "request dma channel[%s] failed\n", channel_names1);
+			ret = -EPROBE_DEFER;
+			goto release_buf0;
 		}
-		dev_err(chip->dev, "dma channel[%d] is NULL\n", SNDRV_PCM_STREAM_PLAYBACK);
+		dev_err(chip->dev, "dma channel[%s] is NULL\n", channel_names1);
 	} else {
-		chip->chan[SNDRV_PCM_STREAM_PLAYBACK] = chan1;
+		chip->chan[SNDRV_PCM_STREAM_CAPTURE] = chan1;
+		period_bytes_max = dma_get_max_seg_size(chan1->device->dev);
+		chip->conv_buf[SNDRV_PCM_STREAM_CAPTURE] = kzalloc(period_bytes_max, GFP_KERNEL);
+		if (!chip->conv_buf[SNDRV_PCM_STREAM_CAPTURE]) {
+			dev_err(chip->dev, "alloc conv buf0 err:%d\n", ret);
+			ret = -ENOMEM;
+			goto release_chan1;
+		}
 	}
 
 	ret = snd_soc_component_initialize(&chip->pcm_component, &esw_pcm_component_driver, chip->dev);
 	if (ret)
-		goto error;
+		goto release_buf1;
 
 	ret = snd_soc_add_component(&chip->pcm_component, NULL, 0);
 	if (ret) {
 		dev_err(chip->dev, "add pcm component failed\n");
-		goto error;
+		goto release_buf1;
 	}
 
 	return 0;
 
-error:
-	dma_release_channel(chip->chan[0]);
+release_buf1:
+	kfree(chip->conv_buf[1]);
+release_chan1:
 	dma_release_channel(chip->chan[1]);
-	return -1;
+release_buf0:
+	kfree(chip->conv_buf[0]);
+release_chan0:
+	dma_release_channel(chip->chan[0]);
+
+	return ret;
 }
\ No newline at end of file
diff --git a/sound/soc/eswin/esw-i2s.c b/sound/soc/eswin/esw-i2s.c
index 9e3648863be60..0d40aee31f871 100755
--- a/sound/soc/eswin/esw-i2s.c
+++ b/sound/soc/eswin/esw-i2s.c
@@ -990,6 +990,8 @@ static int i2s_sof_ringbuffer_init(struct i2s_dev *i2s_drvdata)
 										  DSP_RB_DATA_SZIE, DMA_BIDIRECTIONAL, DMA_ATTR_SKIP_CPU_SYNC);
 	if (dma_mapping_error(i2s_drvdata->chan[1]->device->dev, i2s_drvdata->rb_in_iova)) {
 		dev_err(i2s_drvdata->dev, "dma mapping failed\n");
+		dma_unmap_page_attrs(i2s_drvdata->chan[0]->device->dev, i2s_drvdata->rb_out_iova, DSP_RB_DATA_SZIE,
+							 DMA_BIDIRECTIONAL, DMA_ATTR_SKIP_CPU_SYNC);
 		return -ENOMEM;
 	}
 	dev_dbg(i2s_drvdata->dev, "rb in iova_addr:0x%llx\n", i2s_drvdata->rb_in_iova);
@@ -1135,12 +1137,22 @@ static int i2s_remove(struct platform_device *pdev)
 {
 	struct i2s_dev *i2s_drvdata = dev_get_drvdata(&pdev->dev);
 
-	clk_disable_unprepare(i2s_drvdata->mclk);
+	dev_info(&pdev->dev, "dev name:%s\n", pdev->dev.of_node->name);
 
+	clk_disable_unprepare(i2s_drvdata->mclk);
 	pm_runtime_disable(&pdev->dev);
-
 	audio_proc_module_exit();
 
+	if (of_property_read_bool(pdev->dev.of_node, "dsp_sof_enable")) {
+		dma_unmap_page_attrs(i2s_drvdata->chan[0]->device->dev, i2s_drvdata->rb_out_iova, DSP_RB_DATA_SZIE,
+							DMA_BIDIRECTIONAL, DMA_ATTR_SKIP_CPU_SYNC);
+		dma_unmap_page_attrs(i2s_drvdata->chan[1]->device->dev, i2s_drvdata->rb_in_iova, DSP_RB_DATA_SZIE,
+							DMA_BIDIRECTIONAL, DMA_ATTR_SKIP_CPU_SYNC);
+	} else {
+		kfree(i2s_drvdata->conv_buf[0]);
+		kfree(i2s_drvdata->conv_buf[1]);
+	}
+
 	return 0;
 }
 
diff --git a/sound/soc/sof/eswin/Kconfig b/sound/soc/sof/eswin/Kconfig
index 163263f60d409..273e28dcb67b1 100644
--- a/sound/soc/sof/eswin/Kconfig
+++ b/sound/soc/sof/eswin/Kconfig
@@ -25,6 +25,7 @@ config SND_SOC_SOF_ESWIN_COMMON
 config SND_SOC_SOF_ESWIN
 	tristate "SOF support for eswin"
 	select SND_SOC_SOF_ESWIN_COMMON
+	default n
 	help
 	  This adds support for Sound Open Firmware for ESWIN win2030 platforms.
 	  Say Y if you have such a device.
diff --git a/sound/soc/sof/eswin/Makefile b/sound/soc/sof/eswin/Makefile
index 55436a008eac6..dcb9c540b37b6 100644
--- a/sound/soc/sof/eswin/Makefile
+++ b/sound/soc/sof/eswin/Makefile
@@ -1,6 +1,5 @@
 # SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
 ccflags-y += -I$(srctree)/drivers/soc/eswin/
-ccflags-y += -I$(srctree)/drivers/memory/eswin/es_iommu_rsv/include/linux
 
 snd-sof-eswin-dsp-objs := es-sof-dsp.o es-common.o
 
diff --git a/sound/soc/sof/eswin/es-common.h b/sound/soc/sof/eswin/es-common.h
index d9e1803d4ab2c..415ccb866d74a 100644
--- a/sound/soc/sof/eswin/es-common.h
+++ b/sound/soc/sof/eswin/es-common.h
@@ -24,7 +24,7 @@
 
 #include <linux/clk.h>
 #include <linux/mailbox_client.h>
-#include "es_iommu_rsv.h"
+#include <linux/es_iommu_rsv.h>
 #include <linux/iommu.h>
 #include "eswin-dsp-subsys.h"
 
-- 
2.51.0

